### PR TITLE
Document exception for metrics/classification

### LINKED ISSUE #5434: Document exceptions
## 📚 Documentation
This is only a suggestion, but I find it useful to have what exceptions functions/classes can raise in the docs, so I'd like to suggest adding the `Raises:` section to public functions/classes.

**Progress**
`dirname $(grep -rl "raise " pytorch_lightning)|sort|uniq|awk '{print "- [ ] "$1}'`

- [ ] pytorch_lightning/accelerators (#9558)
- [ ] pytorch_lightning/callbacks
- [ ] pytorch_lightning/callbacks/progress
- [ ] pytorch_lightning/core
- [ ] pytorch_lightning/core/mixins
- [ ] pytorch_lightning/loggers
- [ ] pytorch_lightning/loops
- [ ] pytorch_lightning/loops/dataloader
- [ ] pytorch_lightning/loops/epoch
- [ ] pytorch_lightning/loops/optimization
- [ ] pytorch_lightning/overrides
- [ ] pytorch_lightning/plugins
- [ ] pytorch_lightning/plugins/environments
- [ ] pytorch_lightning/plugins/io
- [ ] pytorch_lightning/plugins/precision
- [ ] pytorch_lightning/plugins/training_type
- [ ] pytorch_lightning/profiler
- [ ] pytorch_lightning/trainer
- [ ] pytorch_lightning/trainer/connectors
- [ ] pytorch_lightning/trainer/connectors/logger_connector
- [ ] pytorch_lightning/tuner
- [ ] pytorch_lightning/utilities

### PR DESCRIPTION
**Pull request by dipam7**
## What does this PR do?

Part of #5434 

- [x] __init__.py
- [x] accuracy.py
- [x] auc.py
- [x] auroc.py
- [x] average_precision.py
- [x] confusion_matrix.py
- [x] f_beta.py
- [x] hamming_distance.py
- [ ] helpers.py (I believe these are just helper functions and not displayed in the documentation so they don't need to document exceptions)
- [x] iou.py
- [x] precision_recall.py
- [x] precision_recall_curve.py
- [x] roc.py
- [x] stat_scores.py  

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by akihironitta — COMMENTED** [review]
@dipam7 Thank you for your contribution! Would you mind adding docs to the following files, too?:
- `pytorch_lightning/metrics/utils.py`
- `pytorch_lightning/metrics/metric.py`
- `pytorch_lightning/metrics/compositional.py`

**Comment by dipam7** [comment]
@akihironitta Sure I'll do those as well

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/metrics/classification/auroc.py [hunk 1]
```diff
@@ -65,6 +65,17 @@ class AUROC(Metric):
...
+    Raises:
+        ValueError:
+            If ``average`` is none of ``None``, ``"macro"`` or ``"weighted"``.
+        ValueError:
+            If max_fpr is not a ``float`` in the range ``0`` < max_fpr <= ``1``.
+        RuntimeError:
+            If ``PyTorch version`` is ``below 1.6`` since max_fpr requires `torch.bucketize`
+            which is not available below 1.6.
+        ValueError:
+            If the mode of data (binary, multi-label, multi-class) changes between batches
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/auroc.py:77]:
```suggestion
            If the mode of data (binary, multi-label, multi-class) changes between batches.
```

File: pytorch_lightning/metrics/classification/f_beta.py [hunk 2]
```diff
@@ -64,6 +64,10 @@ class FBeta(Metric):
         process_group:
             Specify the process group on which synchronization is called. default: None (which selects the entire world)
 
+    Raises:
+        ValueError:
+            If ``average`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"none"``, ``None``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/f_beta.py:69]:
```suggestion
            If ``average`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"none"``, ``None``.
```

File: pytorch_lightning/metrics/classification/precision_recall.py [hunk 3]
```diff
@@ -105,6 +105,10 @@ class Precision(StatScores):
             Callback that performs the allgather operation on the metric state. When ``None``, DDP
             will be used to perform the allgather.
 
+    Raises:
+        ValueError:
+            If ``average`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"samples"``, ``"none"``, ``None``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/precision_recall.py:110]:
```suggestion
            If ``average`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"samples"``, ``"none"``, ``None``.
```

File: pytorch_lightning/metrics/classification/precision_recall.py [hunk 4]
```diff
@@ -256,6 +260,10 @@ class Recall(StatScores):
...
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/precision_recall.py:265]:
```suggestion
            If ``average`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"samples"``, ``"none"``, ``None``.
```

File: pytorch_lightning/metrics/classification/stat_scores.py [hunk 5]
```diff
@@ -101,6 +101,19 @@ class StatScores(Metric):
             Callback that performs the allgather operation on the metric state. When ``None``, DDP
             will be used to perform the allgather.
 
+    Raises:
+        ValueError:
+            If ``threshold`` is not a ``float`` between ``0`` and ``1``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/stat_scores.py:106]:
```suggestion
            If ``threshold`` is not a ``float`` between ``0`` and ``1``.
```

File: pytorch_lightning/metrics/classification/stat_scores.py [hunk 6]
```diff
@@ -101,6 +101,19 @@ class StatScores(Metric):
...
+        ValueError:
+            If ``reduce`` is none of ``"micro"``, ``"macro"`` or ``"samples"``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/stat_scores.py:108]:
```suggestion
            If ``reduce`` is none of ``"micro"``, ``"macro"`` or ``"samples"``.
```

File: pytorch_lightning/metrics/classification/stat_scores.py [hunk 7]
```diff
@@ -101,6 +101,19 @@ class StatScores(Metric):
...
+        ValueError:
+            If ``mdmc_reduce`` is none of ``None``, ``"samplewise"``, ``"global"``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/stat_scores.py:110]:
```suggestion
            If ``mdmc_reduce`` is none of ``None``, ``"samplewise"``, ``"global"``.
```

File: pytorch_lightning/metrics/classification/stat_scores.py [hunk 8]
```diff
@@ -101,6 +101,19 @@ class StatScores(Metric):
...
+        ValueError:
+            If ``reduce`` is set to ``"macro"`` and ``num_classes`` is not provided
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/stat_scores.py:112]:
```suggestion
            If ``reduce`` is set to ``"macro"`` and ``num_classes`` is not provided.
```

File: pytorch_lightning/metrics/classification/auroc.py [hunk 9]
```diff
@@ -65,6 +65,17 @@ class AUROC(Metric):
            Callback that performs the allgather operation on the metric state. When ``None``, DDP
            will be used to perform the allgather
 
+    Raises:
+        ValueError:
+            If ``average`` is none of ``None``, ``"macro"`` or ``"weighted"``.
+        ValueError:
+            If max_fpr is not a ``float`` in the range ``0`` < max_fpr <= ``1``.
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/auroc.py:72]:
```suggestion
            If max_fpr is not a ``float`` in the range ``0`` < ``max_fpr`` <= ``1``.
```

File: pytorch_lightning/metrics/classification/auroc.py [hunk 10]
```diff
@@ -65,6 +65,17 @@ class AUROC(Metric):
...
+        RuntimeError:
+            If ``PyTorch version`` is ``below 1.6`` since max_fpr requires `torch.bucketize`
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/auroc.py:74]:
```suggestion
            If ``PyTorch version`` is ``below 1.6`` since max_fpr requires ``torch.bucketize``
```

File: pytorch_lightning/metrics/classification/stat_scores.py [hunk 11]
```diff
@@ -101,6 +101,19 @@ class StatScores(Metric):
...
+        ValueError:
+            If ``num_classes`` is set
+            and ``ignore_index`` is not in the range ``0`` <= ``ignore_index`` < ``num_classes``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/classification/stat_scores.py:115]:
```suggestion
            and ``ignore_index`` is not in the range ``0`` <= ``ignore_index`` < ``num_classes``.
```

File: pytorch_lightning/metrics/metric.py [hunk 12]
```diff
@@ -120,6 +120,11 @@ def add_state(
             When passing a custom function to ``dist_reduce_fx``, expect the synchronized metric state to follow
             the format discussed in the above note.
 
+        Raises:
+            ValueError:
+                If ``default`` is not a ``tensor`` or an ``empty list.``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:125]:
```suggestion
                If ``default`` is not a ``tensor`` or an ``empty list``.
```

File: pytorch_lightning/metrics/metric.py [hunk 13]
```diff
@@ -120,6 +120,11 @@ def add_state(
...
+            ValueError:
+                If ``dist_reduce_fx`` is not callable or one of ``"mean"``, ``"sum"``, ``"cat"``, ``None``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:127]:
```suggestion
                If ``dist_reduce_fx`` is not callable or one of ``"mean"``, ``"sum"``, ``"cat"``, ``None``.
```

File: pytorch_lightning/metrics/metric.py [hunk 14]
```diff
@@ -528,6 +533,14 @@ class MetricCollection(nn.ModuleDict):
               dict as key for output dict. Use this format if you want to chain
               together multiple of the same metric with different parameters.
 
+    Raises:
+        ValueError:
+            If one of the elements of ``metrics`` is not an instance of ``pl.metrics.Metric``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:538]:
```suggestion
            If one of the elements of ``metrics`` is not an instance of ``pl.metrics.Metric``.
```

File: pytorch_lightning/metrics/metric.py [hunk 15]
```diff
@@ -528,6 +533,14 @@ class MetricCollection(nn.ModuleDict):
...
+        ValueError:
+            If two elements in ``metrics`` have the same ``name``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:540]:
```suggestion
            If two elements in ``metrics`` have the same ``name``.
```

File: pytorch_lightning/metrics/metric.py [hunk 16]
```diff
@@ -528,6 +533,14 @@ class MetricCollection(nn.ModuleDict):
...
+        ValueError:
+            If ``metrics`` is not a ``list``, ``tuple`` or a ``dict``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/metric.py:542]:
```suggestion
            If ``metrics`` is not a ``list``, ``tuple`` or a ``dict``.
```

File: pytorch_lightning/metrics/utils.py [hunk 17]
```diff
@@ -60,6 +60,11 @@ def _input_format_classification_one_hot(
         threshold: float used for thresholding multilabel input
         multilabel: boolean flag indicating if input is multilabel
 
+    Raises:
+        ValueError:
+            If preds and target don't have the same number of dimensions
+            or one additional dimension for preds
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/utils.py:66]:
```suggestion
            or one additional dimension for preds.
```

File: pytorch_lightning/metrics/utils.py [hunk 18]
```diff
@@ -247,6 +252,9 @@ def class_reduce(
             - ``'weighted'``: calculate metrics for each label, and find their weighted mean.
             - ``'none'`` or ``None``: returns calculated metric per class
 
+    Raises:
+        ValueError:
+            If ``class_reduction`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"none"`` or ``None``
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/utils.py:257]:
```suggestion
            If ``class_reduction`` is none of ``"micro"``, ``"macro"``, ``"weighted"``, ``"none"`` or ``None``.
```

File: pytorch_lightning/metrics/utils.py [hunk 19]
```diff
@@ -279,6 +287,10 @@ def _stable_1d_sort(x: torch, N: int = 2049):
     if number of elements are larger than 2048. This function pads the tensors,
     makes the sort and returns the sorted array (with the padding removed)
     See this discussion: https://discuss.pytorch.org/t/is-torch-sort-stable/20714
+
+    Raises:
+        ValueError:
+            If dim of x is greater than 1 since stable sort works with only 1d tensors
```
Comment (SkafteNicki) [inline pytorch_lightning/metrics/utils.py:293]:
```suggestion
            If dim of ``x`` is greater than 1 since stable sort works with only 1d tensors.
```

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/metrics/classification/auroc.py [hunk 20]
```diff
@@ -65,6 +65,17 @@ class AUROC(Metric):
            Callback that performs the allgather operation on the metric state. When ``None``, DDP
            will be used to perform the allgather
 
+    Raises:
+        ValueError:
+            If ``average`` is none of ``None``, ``"macro"`` or ``"weighted"``.
+        ValueError:
+            If max_fpr is not a ``float`` in the range ``0`` < ``max_fpr`` <= ``1``.
```
Comment (akihironitta) [inline pytorch_lightning/metrics/classification/auroc.py:72]: @dipam7 Great addition! Could you follow @rohitgr7's suggestion in https://github.com/PyTorchLightning/pytorch-lightning/pull/6273#discussion_r585467010?
```suggestion
            If ``max_fpr`` is not a ``float`` in the range ``(0, 1]``.
```

File: pytorch_lightning/metrics/utils.py [hunk 21]
```diff
@@ -60,6 +60,11 @@ def _input_format_classification_one_hot(
         threshold: float used for thresholding multilabel input
         multilabel: boolean flag indicating if input is multilabel
 
+    Raises:
+        ValueError:
+            If preds and target don't have the same number of dimensions
+            or one additional dimension for preds.
```
Comment (akihironitta) [inline pytorch_lightning/metrics/utils.py:66]:
```suggestion
            If ``preds`` and ``target`` don't have the same number of dimensions
            or one additional dimension for ``preds``.
```

**Review by ananyahjha93 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
